Multi-Disk auto-recovery.

amosbird opened the pull request:
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- Improvement

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):

Now ReplicatedMergeTree can recover data when some of its disks are broken.

Detailed description / Documentation draft:

~~This PR can be viewed as semi-WIP. It works fine but lacks thorough tests (Also I don't know good ways of constructing CI tests for it).~~

NOTE, if all disks are broken, the instance is still alive but will not recover data after some disks become alive. Manual `SYSTEM REPLICA SYNC` command is needed.

### update

This PR has been used more than half year in production and has recovered hundreds of disks. I'll start making reasonable CI test for it.

### update 2

CI tests are added.

abyss7 commented:
Stress-test has found a heap-after-free and stacktrace is correlated with your changes. Looks like there is an issue somewhere in your code. Will take a closer look.

https://gist.github.com/abyss7/e8dcc7f3ce7e1cc335f5f04507b9661a

amosbird commented:
Ah, I forgot to shutdown `tmp_volume`. Now it should be fine.

alexey-milovidov commented:
Need to fix integration tests.

amosbird commented:
Ubsan error query: `SELECT decrypt('aes-128-gcm', 'text', 'key', 'IV')`

sundy-li commented on programs/server/Server.cpp:
```diff
@@ -147,6 +147,7 @@ namespace
 {
 
 void setupTmpPath(Poco::Logger * log, const std::string & path)
+try
```
wrong function

azat commented on src/Disks/DiskLocal.cpp:
```diff
@@ -317,64 +367,160 @@ SyncGuardPtr DiskLocal::getDirectorySyncGuard(const String & path) const
+    catch (...)
+    {
+        LOG_WARNING(logger, "Cannot create the directory of disk {} ({}).", name, disk_path);
+        return false;
```
I think that it should fail if setupAndCheck failed for the first time, consider the case when the permissions for the directory is incorrect, ignoring such disks does not sounds correct behavior to me, the server should fail I think (otherwise it will hide this error and user may realize that not all disks are used only a year)

amosbird replied:
But it's a common case that broken disks aren't recovered yet and we restart.

azat commented on src/Disks/DiskLocal.cpp:
```diff
@@ -317,64 +367,160 @@ SyncGuardPtr DiskLocal::getDirectorySyncGuard(const String & path) const
+            size_t offset = 0;
+            while (buf_size > 0)
+            {
+                ssize_t rd_size = func(fd, ptr, buf_size, offset);
```
Seems that it is either write or read size?

```suggestion
                ssize_t size = func(fd, ptr, buf_size, offset);
```

azat commented on src/Disks/DiskLocal.cpp:
```diff
@@ -317,64 +367,160 @@ SyncGuardPtr DiskLocal::getDirectorySyncGuard(const String & path) const
+    String file_path = disk_path + ".recover";
+    const char * fpath = file_path.c_str();
+    char errmsg[64];
+    if (::access(fpath, F_OK) == 0)
```
Looks like worth adding some `SYSTEM` query for this?
And it seems that it does not do any "recovering", so maybe worth renaming to "rescan" or similar?

amosbird replied:
Luckily we have a `system restart disk` after https://github.com/ClickHouse/ClickHouse/pull/23429

amosbird replied:
Done with `system restart disk`

azat commented on src/Interpreters/Context.cpp:
```diff
@@ -459,13 +459,23 @@ struct ContextSharedPart
+            if (merge_tree_disk_selector)
+                const_cast<DiskSelector &>(*merge_tree_disk_selector).shutdown();
+
+            /// Also needs to shutdown tmp volume
```
Can you add a comment why?

azat commented on src/Storages/MergeTree/IMergeTreeDataPart.cpp:
```diff
@@ -1082,35 +1082,50 @@ void IMergeTreeDataPart::renameTo(const String & new_relative_path, bool remove_
+        {
+            // Don't throw when the destination is to the detached folder. It might be able to
+            // recover in some cases, such as fetching parts into multi-disks while some of the
+            // disks are broken.
```
But it will retry to detach to the same disk on and on, right, so it will never succeed, no?

amosbird replied:
It won't retry. The parts are by definition temporary but somehow more valuable. That's why we keep it under the detached folder.

azat commented on src/Storages/MergeTree/MergeTreeData.cpp:
```diff
@@ -937,6 +947,7 @@ void MergeTreeData::loadDataParts(bool skip_sanity_checks)
 
     std::mutex mutex;
 
+    DataPartsVector duplicate_parts_to_remove;
```
Worth putting a comment why this is possible? (because some of disk may be broken and then can be recovered)

amosbird replied:
I feel that it's evident. (BTW there might be other cases to encounter duplicated parts)

azat reviewed (commented):
So now it expose `isBroken` API, have you considered marking parts as broken instead of marking the whole disk as broken?

For example if you have `partition by toYYYYMM(date)` and you have lots of stale data which you don't read everyday, after this patch all of those stale data will be downloaded from replicas, right?

azat commented on src/Disks/DiskLocal.cpp:
```diff
@@ -317,64 +367,160 @@ SyncGuardPtr DiskLocal::getDirectorySyncGuard(const String & path) const
+    bool read_write_ok = true;
+    if (disk_checker)
+    {
+        auto rw_test = [fd](char * buf, auto && func)
```
Using one code for the read and write looks error prone.

And seems that there is error even now, consider the case when the disk is full and you cannot even write 4K (BUF_SIZE), in this case you will mark the disk as broken, while you should not.

amosbird replied:
I've added `errno==ENOSPC` check.

alexey-milovidov replied:
If the disk is full - it is non-writable. The same if filesystem is mounted in readonly after error. Why these cases are different?

alexey-milovidov replied:
And if disk is non-writeable for whatever reason, we don't have to repair parts to another disk.

azat replied:
>If the disk is full - it is non-writable

It can be temporary full, i.e. due to some temporary files (that can be created by ClickHouse itself)

amosbird commented:
> and you have lots of stale data which you don't read everyday, after this patch all of those stale data will be downloaded from replicas, right?

But that's intentional. We don't want to have broken parts discovered by user queries.

amosbird commented:
Test failures seem unrelated.

alexey-milovidov commented on programs/server/Server.cpp:
```diff
@@ -213,6 +214,10 @@ void setupTmpPath(Poco::Logger * log, const std::string & path)
 }
+catch (...)
+{
+    DB::tryLogCurrentException(log, "Caught exception while setup temporary path: " + path);
```
Add a comment:

"It is ok to skip this exception as cleaning old temporary files is not necessary."

Otherwise catching exception looks suspicious.

alexey-milovidov commented on src/Disks/DiskLocal.cpp:
```diff
@@ -55,9 +58,6 @@ static void loadDiskLocalConfig(const String & name,
             throw Exception("Disk path must end with /. Disk " + name, ErrorCodes::UNKNOWN_ELEMENT_IN_CONFIG);
     }
 
-    if (!FS::canRead(path) || !FS::canWrite(path))
```
I missed this change... does it prevent attaching tables and doing SELECTs from readonly fs? We don't need it.

amosbird replied:
Do you mean we should skip the check of `FS::canWrite(disk_path)`? I guess we can mark it read only and also set the available space to zero to prevent any write intentions.

alexey-milovidov replied:
Yes, that will make sense.

alexey-milovidov commented on src/Disks/DiskLocal.cpp:
```diff
@@ -91,13 +91,50 @@ class DiskLocalReservation : public IReservation
+    {
+        if (i != 0)
+        {
+            throw Exception("Can't use i != 0 with single disk reservation", ErrorCodes::INCORRECT_DISK_INDEX);
```
Maybe logical error - to act as assert?

alexey-milovidov commented on src/Disks/DiskLocal.cpp:
```diff
@@ -380,61 +426,155 @@ void DiskLocal::applyNewSettings(const Poco::Util::AbstractConfiguration & confi
+        return false;
+    }
+
+    String file_path = disk_path + ".check";
```
We need to use std::filesystem::operator/ to concatenate paths.

azat commented on src/Disks/DiskLocal.cpp:
```diff
@@ -380,61 +426,155 @@ void DiskLocal::applyNewSettings(const Poco::Util::AbstractConfiguration & confi
+                if (size <= 0)
+                {
+                    if (errno != ENOSPC)
+                        return false;
```
So now it will endlessly retry in case of ENOSPC until space will be freed?

amosbird replied:
This code will be refactored. ENOSPC is not an exception, and it will not get write intention anyway.

azat commented on src/Disks/DiskLocalCheckThread.cpp:
```diff
@@ -0,0 +1,60 @@
+    if (disk->setupAndCheck())
+    {
+        if (disk->broken)
+            LOG_INFO(log, "Disk seems to be fine. It can be recovered using `SYSTEM RESTART DISK {}`", disk->getName());
```
```suggestion
            LOG_INFO(log, "Disk seems to be broken. It can be recovered using `SYSTEM RESTART DISK {}`", disk->getName());
```

?

amosbird replied:
It means the disk is ok but somehow marks broken. It can be recovered manually.

alexey-milovidov commented:
GitHub is notoriously unstable in the last day, so CI is delayed.

@mergifyio update

alexey-milovidov commented:
Looks like it's almost ready for merge but some PR comments need to be addressed.

amosbird committed:
DiskLocal checker

Add DiskLocal checker so that ReplicatedMergeTree can recover data when some of its disks are broken.

amosbird commented:
I've refactored the code and hopefully addressed all the review comments.

alexey-milovidov commented:
@Mergifyio update

mergify[bot] committed:
Merge branch 'master' into mdha

azat commented on src/Disks/DiskLocal.cpp:
```diff
@@ -409,67 +458,212 @@ void DiskLocal::applyNewSettings(const Poco::Util::AbstractConfiguration & confi
+                throw Exception(ErrorCodes::BAD_ARGUMENTS, "Disk {} and disk {} cannot have the same path ({})", name, disk_name, path);
+
+        std::shared_ptr<IDisk> disk
+            = std::make_shared<DiskLocal>(name, path, keep_free_space_bytes, context, config.getUInt("local_disk_check_period_ms", 0));
```
@amosbird is default value for `local_disk_check_period_ms` sane? This will cause lots of excessive IO, right? (note, that I saw this line in two hunks)

azat replied:
Oh, I see, 0 means it is disabled. Sorry for the noise.

azat commented on src/Disks/DiskLocalCheckThread.cpp:
```diff
@@ -0,0 +1,70 @@
+    bool can_write = disk->canWrite();
+    if (can_read)
+    {
+        if (disk->broken)
```
Why not check for `can_write` too?

zhouerlin commented:
When the disk is abnormal (chmod 000 /disk1), the getFileNamesWithoutChecksums interface cannot access the disk in the checkPart func. As a result, the abnormal part cannot be recovered from the replica, Or the test scenario is not supported ？


2022.06.10 10:41:59.647393 [ 27394 ] {} <Error> test.oper_log (ReplicatedMergeTreePartCheckThread)[150] : void DB::ReplicatedMergeTreePartCheckThread::run(): std::exception. Code: 1001, type: std::__1::__fs::filesystem::filesystem_error, e.what() = filesystem error: in posix_stat: failed to determine attributes for the specified path: Permission denied [ /disks/pshard/data1/store/8be/8bec34f4-a56d-4c1a-8bec-34f4a56d0c1a/202004_0_9_2/default_compression_codec.txt], Stack trace (when copying this message, always include the lines below):

0.  /workspace/ClickHouse-21.8.5.7/contrib/libcxx/include/exception:133: std::exception::capture() @ 0x11eebc41 in  /workspace/ClickHouse-21.8.5.7/build_debug/programs/clickhouse
1.  /workspace/ClickHouse-21.8.5.7/contrib/libcxx/include/exception:111: std::runtime_error::runtime_error(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) @ 0x2c7afb77 in  /workspace/ClickHouse-21.8.5.7/build_debug/programs/clickhouse
2.  /workspace/ClickHouse-21.8.5.7/contrib/libcxx/src/system_error.cpp:244: std::__1::system_error::system_error(std::__1::error_code, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) @ 0x2c7c443e in  /workspace/ClickHouse-21.8.5.7/build_debug/programs/clickhouse
3.  /workspace/ClickHouse-21.8.5.7/contrib/libcxx/include/filesystem:1614: void std::__1::__fs::filesystem::detail::(anonymous namespace)::ErrorHandler<void>::report<>(std::__1::error_code const&, char const*) const @ 0x2c6f57a7 in  /workspace/ClickHouse-21.8.5.7/build_debug/programs/clickhouse
4.  /workspace/ClickHouse-21.8.5.7/contrib/libcxx/src/filesystem/operations.cpp:468: std::__1::__fs::filesystem::detail::(anonymous namespace)::create_file_status(std::__1::error_code&, std::__1::__fs::filesystem::path const&, stat const&, std::__1::error_code*) @ 0x2c6e0c7a in  /workspace/ClickHouse-21.8.5.7/build_debug/programs/clickhouse
5.  /workspace/ClickHouse-21.8.5.7/contrib/libcxx/src/filesystem/operations.cpp:501: std::__1::__fs::filesystem::detail::(anonymous namespace)::posix_stat(std::__1::__fs::filesystem::path const&, stat&, std::__1::error_code*) @ 0x2c6e0f47 in  /workspace/ClickHouse-21.8.5.7/build_debug/programs/clickhouse
6.  /workspace/ClickHouse-21.8.5.7/contrib/libcxx/src/filesystem/operations.cpp:506: std::__1::__fs::filesystem::detail::(anonymous namespace)::posix_stat(std::__1::__fs::filesystem::path const&, std::__1::error_code*) @ 0x2c6e0fa4 in  /workspace/ClickHouse-21.8.5.7/build_debug/programs/clickhouse
7.  /workspace/ClickHouse-21.8.5.7/contrib/libcxx/src/filesystem/operations.cpp:1296: std::__1::__fs::filesystem::__status(std::__1::__fs::filesystem::path const&, std::__1::error_code*) @ 0x2c6e8b2b in  /workspace/ClickHouse-21.8.5.7/build_debug/programs/clickhouse
8.  /workspace/ClickHouse-21.8.5.7/contrib/libcxx/include/filesystem:1899: DB::DiskLocal::exists(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) const @ 0x20c7f4ab in  /workspace/ClickHouse-21.8.5.7/build_debug/programs/clickhouse
9.  /workspace/ClickHouse-21.8.5.7/src/Storages/MergeTree/IMergeTreeDataPart.cpp:688: DB::IMergeTreeDataPart::getFileNamesWithoutChecksums() const @ 0x230f43e0 in  /workspace/ClickHouse-21.8.5.7/build_debug/programs/clickhouse
10.  /workspace/ClickHouse-21.8.5.7/src/Storages/MergeTree/checkDataPart.cpp:293: DB::checkDataPart(std::__1::shared_ptr<DB::IMergeTreeDataPart const>, bool, std::__1::function<bool ()>) @ 0x2378f161 in  /workspace/ClickHouse-21.8.5.7/build_debug/programs/clickhouse
11.  /workspace/ClickHouse-21.8.5.7/src/Storages/MergeTree/ReplicatedMergeTreePartCheckThread.cpp:303: DB::ReplicatedMergeTreePartCheckThread::checkPart(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) @ 0x236d08aa in  /workspace/ClickHouse-21.8.5.7/build_debug/programs/clickhouse
12.  /workspace/ClickHouse-21.8.5.7/src/Storages/MergeTree/ReplicatedMergeTreePartCheckThread.cpp:414: DB::ReplicatedMergeTreePartCheckThread::run() @ 0x236d45fb in  /workspace/ClickHouse-21.8.5.7/build_debug/programs/clickhouse
13.  /workspace/ClickHouse-21.8.5.7/src/Storages/MergeTree/ReplicatedMergeTreePartCheckThread.cpp:32: DB::ReplicatedMergeTreePartCheckThread::ReplicatedMergeTreePartCheckThread(DB::StorageReplicatedMergeTree&)::'lambda'()::operator()() const @ 0x236c527b in  /workspace/ClickHouse-21.8.5.7/build_debug/programs/clickhouse
14.  /workspace/ClickHouse-21.8.5.7/contrib/libcxx/include/type_traits:3676: void std::__1::__invoke_void_return_wrapper<void>::__call<DB::ReplicatedMergeTreePartCheckThread::ReplicatedMergeTreePartCheckThread(DB::StorageReplicatedMergeTree&)::'lambda'()&>(DB::ReplicatedMergeTreePartCheckThread::ReplicatedMergeTreePartCheckThread(DB::StorageReplicatedMergeTree&)::'lambda'()&) @ 0x236d5470 in  /workspace/ClickHouse-21.8.5.7/build_debug/programs/clickhouse
15.  /workspace/ClickHouse-21.8.5.7/contrib/libcxx/include/functional:1608: void std::__1::__function::__policy_invoker<void ()>::__call_impl<std::__1::__function::__default_alloc_func<DB::ReplicatedMergeTreePartCheckThread::ReplicatedMergeTreePartCheckThread(DB::StorageReplicatedMergeTree&)::'lambda'(), void ()> >(std::__1::__function::__policy_storage const*) @ 0x236d53e8 in  /workspace/ClickHouse-21.8.5.7/build_debug/programs/clickhouse
16.  /workspace/ClickHouse-21.8.5.7/contrib/libcxx/include/functional:2222: std::__1::function<void ()>::operator()() const @ 0x11f3d1ad in  /workspace/ClickHouse-21.8.5.7/build_debug/programs/clickhouse
17.  /workspace/ClickHouse-21.8.5.7/src/Core/BackgroundSchedulePool.cpp:106: DB::BackgroundSchedulePoolTaskInfo::execute() @ 0x20c9364e in  /workspace/ClickHouse-21.8.5.7/build_debug/programs/clickhouse
18.  /workspace/ClickHouse-21.8.5.7/src/Core/BackgroundSchedulePool.cpp:19: DB::TaskNotification::execute() @ 0x20c9aee7 in  /workspace/ClickHouse-21.8.5.7/build_debug/programs/clickhouse
19.  /workspace/ClickHouse-21.8.5.7/src/Core/BackgroundSchedulePool.cpp:262: DB::BackgroundSchedulePool::threadFunction() @ 0x20c971aa in  /workspace/ClickHouse-21.8.5.7/build_debug/programs/clickhouse
20.  /workspace/ClickHouse-21.8.5.7/src/Core/BackgroundSchedulePool.cpp:161: DB::BackgroundSchedulePool::BackgroundSchedulePool(unsigned long, unsigned long, char const*)::'lambda'()::operator()() const @ 0x20c9413b in  /workspace/ClickHouse-21.8.5.7/build_debug/programs/clickhouse
21.  /workspace/ClickHouse-21.8.5.7/contrib/libcxx/include/type_traits:3682: ThreadFromGlobalPool::ThreadFromGlobalPool<DB::BackgroundSchedulePool::BackgroundSchedulePool(unsigned long, unsigned long, char const*)::'lambda'()>(DB::BackgroundSchedulePool::BackgroundSchedulePool(unsigned long, unsigned long, char const*)::'lambda'()&&)::'lambda'()::operator()() @ 0x20c9827e in  /workspace/ClickHouse-21.8.5.7/build_debug/programs/clickhouse
22.  /workspace/ClickHouse-21.8.5.7/contrib/libcxx/include/type_traits:3676: void std::__1::__invoke_void_return_wrapper<void>::__call<ThreadFromGlobalPool::ThreadFromGlobalPool<DB::BackgroundSchedulePool::BackgroundSchedulePool(unsigned long, unsigned long, char const*)::'lambda'()>(DB::BackgroundSchedulePool::BackgroundSchedulePool(unsigned long, unsigned long, char const*)::'lambda'()&&)::'lambda'()&>(DB::BackgroundSchedulePool::BackgroundSchedulePool(unsigned long, unsigned long, char const*)::'lambda'()&&...) @ 0x20c9aa50 in  /workspace/ClickHouse-21.8.5.7/build_debug/programs/clickhouse
23.  /workspace/ClickHouse-21.8.5.7/contrib/libcxx/include/functional:1608: void std::__1::__function::__policy_invoker<void ()>::__call_impl<std::__1::__function::__default_alloc_func<ThreadFromGlobalPool::ThreadFromGlobalPool<DB::BackgroundSchedulePool::BackgroundSchedulePool(unsigned long, unsigned long, char const*)::'lambda'()>(DB::BackgroundSchedulePool::BackgroundSchedulePool(unsigned long, unsigned long, char const*)::'lambda'()&&)::'lambda'(), void ()> >(std::__1::__function::__policy_storage const*) @ 0x20c9a4cb in  /workspace/ClickHouse-21.8.5.7/build_debug/programs/clickhouse
24.  /workspace/ClickHouse-21.8.5.7/contrib/libcxx/include/functional:2222: std::__1::function<void ()>::operator()() const @ 0x11f3d1ad in  /workspace/ClickHouse-21.8.5.7/build_debug/programs/clickhouse
25.  /workspace/ClickHouse-21.8.5.7/src/Common/ThreadPool.cpp:269: ThreadPoolImpl<std::__1::thread>::worker(std::__1::__list_iterator<std::__1::thread, void*>) @ 0x11f75239 in  /workspace/ClickHouse-21.8.5.7/build_debug/programs/clickhouse
26.  /workspace/ClickHouse-21.8.5.7/src/Common/ThreadPool.cpp:136: void ThreadPoolImpl<std::__1::thread>::scheduleImpl<void>(std::__1::function<void ()>, int, std::__1::optional<unsigned long>)::'lambda1'()::operator()() const @ 0x11f77ee6 in  /workspace/ClickHouse-21.8.5.7/build_debug/programs/clickhouse
27.  /workspace/ClickHouse-21.8.5.7/contrib/libcxx/include/type_traits:3676: void* std::__1::__thread_proxy<std::__1::tuple<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct> >, void ThreadPoolImpl<std::__1::thread>::scheduleImpl<void>(std::__1::function<void ()>, int, std::__1::optional<unsigned long>)::'lambda1'()> >(void*) @ 0x11f8459b in  /workspace/ClickHouse-21.8.5.7/build_debug/programs/clickhouse
28. start_thread @ 0x7df5 in /usr/lib64/libpthread-2.17.so
29. __clone @ 0x10248d in /usr/lib64/libc-2.17.so

azat commented on src/Disks/DiskLocal.cpp:
```diff
@@ -409,67 +458,212 @@ void DiskLocal::applyNewSettings(const Poco::Util::AbstractConfiguration & confi
+    try
+    {
+        if (!FS::canRead(disk_path))
+            throw Exception(ErrorCodes::PATH_ACCESS_DENIED, "There is no read access to disk {} ({}).", name, disk_path);
```
@amosbird should it `return false` here instead?

Otherwise this method never returns `false` and return value is simply useless, and the line `readonly = !setup()` does not makes a lot of sense.

azat replied:
Something like this https://github.com/ClickHouse/ClickHouse/pull/43143/commits/ba6b5bd7605745b1db247646f7f2e5b089516fcf

amosbird replied:
IIRC, it's because `FS::canRead` might also throw exception, so instead of returning false we throw exception instead.

> return value is simply useless, 

Indeed. The function signature should be changed to `void setup()`.

azat replied:
https://github.com/ClickHouse/ClickHouse/blob/0b4e643c2764543911b0b4c0c3745846af05742e/src/Common/filesystemHelpers.cpp#L313-L326

Looking at `FS::canRead` and it seems that it makes sense to do something like this:

```c++
try {
  readonly = !FS::canRead();
} catch (...) {
  readonly = true;
}
```

amosbird replied:
But `readonly` is not `broken`.

amosbird replied:
>  @amosbird should it return false here instead?

Ah yes, it should.
